src/native/minipal/utf8.c

```diff
@@ -0,0 +1,2198 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include <minipal/utf8.h>
+
+#include <errno.h>
+#include <limits.h>
+#include <string.h>
+#include <assert.h>
+
+#define FASTLOOP
+
+#define HIGH_SURROGATE_START 0xd800
+#define HIGH_SURROGATE_END 0xdbff
+#define LOW_SURROGATE_START 0xdc00
+#define LOW_SURROGATE_END 0xdfff
+
+// Test if the wide character is a high surrogate
+static bool IsHighSurrogate(const CHAR16_T c)
+{
+    return (c & 0xFC00) == HIGH_SURROGATE_START;
+}
+
+// Test if the wide character is a low surrogate
+static bool IsLowSurrogate(const CHAR16_T c)
+{
+    return (c & 0xFC00) == LOW_SURROGATE_START;
+}
+
+// Test if the wide character is a surrogate half
+static bool IsSurrogate(const CHAR16_T c)
+{
+    return (c & 0xF800) == HIGH_SURROGATE_START;
+}
+
+static size_t minipal_wcslen(const CHAR16_T* str)
+{
+    size_t nChar = 0;
+    while (*str++) nChar++;
+    return nChar;
+}
+
+#define ContractAssert(cond)                     \
```

Comment:
Should these be just a regular debug-only asserts? I think that the original code had a bug where these were not debug-only.

Reply:
Original implementation was throwing exception and caller was catching it and mapping it to downstream errno. Now we are directly setting the errno.

This is `MB_ERR_INVALID_CHARS` feature. In coreclr `MAKE_WIDEPTR_FROMUTF8N_NOTHROW` is the only macro using it (rest of the callers are dead code..). It is used by `custattr_emit.cpp` and there are some (non-realistic) tests which rely on it: https://github.com/dotnet/runtime/blob/25a4e045aa975656a96e3edd8380093177fe841d/src/libraries/System.Reflection.Emit/tests/TypeBuilder/TypeBuilderDefineEvent.cs#L23-L25

Definitely room for cleanups here if we remove `MB_ERR_INVALID_CHARS`. Can be done as a follow-up.

Reply:
This was based on the original managed Utf8 encoding implementation: https://github.com/dotnet/coreclr/blob/release/1.0.0/src/mscorlib/src/System/Text/UTF8Encoding.cs `Contract.Assert` was debug-only in the original managed version.

It sounds like some other checks got converted to Contract Assert somehow. What is the error path that is important for these tests?

Reply:
Converted to assert and cleaned up the unused macros from utilcode.